### PR TITLE
Implementation of MC bending sampling

### PR DESCRIPTION
**Pull request by fgrunewald**
This code implements a local bending stiffness to the polymers. The idea is that one can set a bending constant for any triplet of residues. Subsequently, during the RW protocol the angle if a trail is accepted or rejected according to a MC-like procedure. First, the normalized probability of an angle is computed using: 

$p(\theta) = \huge \frac{e^{\frac{lp \times \theta}{180}}}{\frac{180 \times e^{lp}-1}{lp}}$

This probability distribution is about equal for low values of lp and has an increased probability at higher angles with increasing lp. Note lp is a dimensionless bending constant not per se the persistence length. Once the probability is computed we check if the value is higher than the previous bending probability. If yes the move is accepted and if not we compare it to the probability of a uniform sampling of the distribution. 

So far bending constant can be given as tuples of three residue names. This means local bending stiffness can be accounted for. We could therefore use it to really fine-tune the DNA parameters. However, it also means the bending term is non-symmetric which means for combinations of many different monomers this might become cumbersome. Ideally, we could set a default value for all monomers. 

Points to discuss:
- Do we want to be able to set one default value for all triplets?
- Is there a better probability distribution?
- Do we want to see if there is an approximate relation between lp reported and lp set? 
- Allow bending in combination with restraints? Should probably be tested?
- do we want to add a temperature factor?

@jan-stevens let me know your thoughts. 

I did some testing of this implementation and in general works quite well to reproduce the expected end-to-end distance by setting an appropriately large bending constant. Any relation to the experimental value of lp is, however, to me not obvious. Probably the estimated volume of the residue itself contributes to the stiffness to some extent so one potentially has to determine a qualitative bending constant by trail and error. For DNA, we can simply figure it out and report. The main limitation of this implementation is that the distributions are not as peaked for cases where `contour-length >~ lp`. In those cases the WCM distributions are probably not too good either.

### CODE REVIEW

**Comment by fgrunewald** [comment]
some observation:

stiff polymers where `contour-length >~ lp` need a low number of rewind steps (e.g. setting `-rw 1`). For single polymers this helps to get relatively straight chain conformations.

**Comment by fgrunewald** [comment]
summoning @pckroon for a technical code review

File: polyply/src/random_walk.py [hunk 1]
```diff
@@ -356,10 +395,9 @@ def _random_walk(self, meta_molecule):
         count = 0
         path = list(meta_molecule.search_tree.edges)
         step_count = 0
-
         while step_count < len(path):
             prev_node, current_node = path[step_count]
-
+            print(step_count)
```
Comment (pckroon) [inline polyply/src/random_walk.py:400]: Leftover

File: polyply/tests/test_nb_engine.py [hunk 2]
```diff
@@ -172,6 +172,31 @@ def test_update_positions_in_molecules(topology):
...
+    # we add two fixed positions to the first molecule
+    positions = [np.array([0., 0., 0,]), np.array([0., 0., 1.])]
+    for node, position in zip([0, 1], positions):
+        topology.molecules[0].nodes[node]["position"] = position
+
+    nb_engine =  NonBondEngine.from_topology(topology.molecules,
+                                             topology,
+                                             box=np.array([10., 10., 10.]))
+    prob = nb_engine.compute_bending_probability(lp, point, 0, 1, 0)
+    assert np.isclose(prob, ref_prob, atol=10**-5)
```
Comment (pckroon) [inline polyply/tests/test_nb_engine.py:198]:
```suggestion
    assert prob == pytest.approx(ref_prob, abs=10**-5)
```

File: polyply/tests/test_random_walk.py [hunk 3]
```diff
@@ -188,6 +189,63 @@ def test_is_overlap(nonbond_matrix, molecule, new_point, result):
...
+     10,
+     True
+     ),
+    # medium angle, prev prob high, unifrom prob low
+    (2,
+     np.array([1., 1.57, 1.74]),
+     1.,
+     10,
+     True
+     ),
```
Comment (pckroon) [inline polyply/tests/test_random_walk.py:234]: I'd prefer each of these cases on 1 line

**Review by pckroon — CHANGES_REQUESTED** [review]
Some small comments, mostly superficial

File: polyply/src/nonbond_engine.py [hunk 4]
```diff
@@ -292,6 +302,42 @@ def compute_force_point(self, point, mol_idx, node, exclude=[], potential="LJ"):
...
+        Compute probability of an angle between three points (`point`, `node_b`,
+        `node_c`) according to a simple decay function modulated by a bending
+        constant `lp`. If lp is large the distribution increases with the highest
+        probability being at 180 degrees (i.e. more straight angles), whilst if lp
+        is low the angles occur with almost equal probability.
+
+        Parameters
+        ----------
+        lp: float
+            bending constant
```
[the remark this answers is not in the document]
Comment (fgrunewald) [reply polyply/src/nonbond_engine.py:316]: nope it's unitless; it's not like a physically meaningful constant but rather if it's low the polymer if flexible if it's high it's stiff. The actual bending is also determined by the volume and force exclusion criteria used in the RW

File: polyply/src/nonbond_engine.py [hunk 5]
```diff
@@ -292,6 +302,42 @@ def compute_force_point(self, point, mol_idx, node, exclude=[], potential="LJ"):
...
+            the probability
+        """
+        # get the missing postions
+        B_pos = self.get_point(mol_idx, node_b)
+        C_pos = self.get_point(mol_idx, node_c)
+        # compute angle
+        ang_val = angle(point, B_pos, C_pos)
+        # compute probability
+        # denominator is normalization
+        prob = np.exp(lp*ang_val/180)/(180*(np.exp(lp)-1)/lp)
```
[the remark this answers is not in the document]
Comment (fgrunewald) [reply polyply/src/nonbond_engine.py:338]: the denominator is the normalisation, which is why it is written a bit funny; I'll put a comment there and keep it.

File: polyply/src/random_walk.py [hunk 6]
```diff
@@ -279,6 +282,41 @@ def checks_milestones(self, current_node, current_position, fudge=0.7):
...
+                                                                       b_node,
+                                                                       c_node)
+                if self.prev_prob < prob:
+                    self.prev_prob = prob
+                    return True
+
+                # compute test probability from uniform sampling of prob function
+                test_prob = random.uniform(np.exp(lp*1/180)/(180*(np.exp(lp)-1)/lp),
+                                           np.exp(lp*179/180)/(180*(np.exp(lp)-1)/lp))
+                if prob > test_prob:
```
[the remark this answers is not in the document]
Comment (fgrunewald) [reply polyply/src/random_walk.py:312]: Yes, they are. In the first case, we test if the new state is more likely than the previous one. If yes, we accept it right away, and if not we test if the state is more likely than a randomly pulled probability from a uniform distribution.

File: bin/polyply [hunk 7]
```diff
@@ -163,6 +163,9 @@ def main(): # pylint: disable=too-many-locals,too-many-statements
     random_walk_group.add_argument('-nr', dest='nrewind', type=int, default=5,
                                    help=('number of residues to trace back '
                                          ' when RW fails in first attempt'))
+    random_walk_group.add_argument('-mb', dest='max_bend', type=float, default=0,
+                                   help=('persistence length determining max bending '
+                                         'the higher the more straight the chain.'))
```
Comment (fgrunewald) [inline bin/polyply:168]: remove this option

**Review by pckroon — DISMISSED** [review]
LGTM, assuming you addressed all my previous comments

**Review by pckroon — APPROVED** [review]

**Pull request merged**